src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs

```diff
@@ -10,98 +10,123 @@ namespace Microsoft.Extensions.Primitives
     [DebuggerDisplay("Value = {_value}")]
     public struct InplaceStringBuilder
     {
-        private int _capacity;
         private int _offset;
-        private bool _writing;
         private string _value;
 
         public InplaceStringBuilder(int capacity) : this()
         {
-            _capacity = capacity;
+            if (capacity < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            }
+
+            _value = new string('\0', capacity);
         }
 
         public int Capacity
         {
-            get { return _capacity; }
+            get => _value?.Length ?? 0;
             set
             {
                 if (value < 0)
                 {
-                    throw new ArgumentOutOfRangeException(nameof(value));
+                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.value);
                 }
-                if (_writing)
+
+                // _offset > 0 indicates writing state
+                if (_offset > 0)
                 {
-                    throw new InvalidOperationException("Cannot change capacity after write started.");
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.Capacity_CannotChangeAfterWriteStarted);
```

Comment:
Why did you move this exception message, but not the other to the resources?

Reply:
For me the `ThrowHelper` is a reusable piece of code.

`ThrowValidationError` is special to the `InplaceStringBuilder` and accesses `Capacity`, thats way this is kept local.

`ThrowNotEnoughCapacity` and `ThrowCapacityNotUsed` are also specialized for the `InplaceStringBuilder` with access to local data. Plus the string is built at runtime and not so easy to represent as enum -- though still possible.

Or did you mean different exception messages?

Reply:
I mean exception message moved to resources.

Reply:
Sorry, I don't know which ones you mean. Either they have `ExceptionArgument` or the string is built at runtime. Can you tell me the line-no?

Reply:
Oh, I see, it's because it's built at runtime. Let's go with the ThrowHelper pattern corefx uses and put exceptions there and in resources (https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/ThrowHelper.cs)

Reply:
A similar pattern as the linked one is already in use in this `ThrowHelper`, except support for older runtimes, that don't have the no-inline behavior for method that don't return.

I'll move these exceptions to the ThrowHelper. The move to the resources is also good, so no string-resources are in the code directly.